atomic_reactor/plugins/input_osv3.py

```diff
@@ -186,3 +187,72 @@ def run(self):
     @classmethod
     def is_autousable(cls):
         return 'BUILD' in os.environ and 'SOURCE_URI' in os.environ and 'OUTPUT_IMAGE' in os.environ
+
+
+class OSv3SourceContainerInputPlugin(InputPlugin):
+    """Input plugin for building source container images"""
+    key = "osv3_source_container"
+
+    def __init__(self, **kwargs):
+        """
+        constructor
+        """
+        # call parent constructor
+        super(OSv3SourceContainerInputPlugin, self).__init__(**kwargs)
+        self.target_registry = None
+        self.reactor_env = None
+        self.plugins_json = None
+
+    def assert_koji_integration(self):
+        """Fail when koji integration is nto configured as koji is integral
+        part of source containers feature"""
+        koji_map = self.reactor_env.get('koji', {})
+        if not koji_map.get('hub_url'):
+            raise RuntimeError(
+                "Koji-hub URL is not configured. Source container image "
+                "builds require enabled koji integration"
+            )
+
+    def run(self):
+        build_json = get_build_json()
+        image = os.environ['OUTPUT_IMAGE']
+        self.target_registry = os.environ.get('OUTPUT_REGISTRY', None)
+
+        user_params = os.environ['USER_PARAMS']
+        user_data = json.loads(user_params)  # TODO validate user params against schema
+        arrangement_version = user_data.get('arrangement_version', None)
+        self.plugins_json = get_plugins_with_user_data(user_params, user_data)
+        # if we get the USER_PARAMS, we'd better get the REACTOR_CONFIG too
+        reactor_config_map = os.environ['REACTOR_CONFIG']
+        self.reactor_env = read_yaml(reactor_config_map, 'schemas/config.json')
+
+        if arrangement_version and arrangement_version <= 5:
+            raise ValueError('arrangement_version <= 5 is no longer supported')
+
+        self.plugins_json = json.loads(self.plugins_json)
+        # validate json before performing any changes
+        read_yaml(json.dumps(self.plugins_json), 'schemas/plugins.json')
```

Comment:
I know this is copied from elsewhere, but maybe this is better:
```
# validate json before performing any changes
read_yaml(self.plugins_json, 'schemas/plugins.json')
self.plugins_json = json.loads(self.plugins_json)
```

Reply:
reworked